### PR TITLE
Shared C codegen

### PR DESCRIPTION
**Pull request by madman-bob**
Expose the RefC code for compiling CC object files and executables, and add the option to compile them as libraries.

This should result in no behavioural changes to the RefC backend. I'm exposing these functions to enable other backends to make use of them, such as Python.

I wasn't able to run `make install-api` on `master`, so this branch is instead based off the most recent tag (`v0.3.0`). From a quick glance, it appears the only merge issues are a matter of `import`s, but I'm not able verify if any particular solution is correct.

### CODE REVIEW

File: src/Compiler/RefC/CC.idr [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+     dirs <- getDirs
+
+     let libraryFlag = if asLibrary then "-fpic " else ""
+
+     let runccobj = cc ++ " -c " ++ libraryFlag ++ sourceFile ++
+                       " -o " ++ objectFile ++ " " ++
+                       "-I" ++ fullprefix_dir dirs "refc " ++
+                       "-I" ++ fullprefix_dir dirs "include"
+
+     coreLift $ putStrLn runccobj
```
Comment (gallais) [inline src/Compiler/RefC/CC.idr:41]: Should this `putStrLn` be a call to `log "compile.refc"` instead?
(You'll need to import `Core.Context.Log` to get access to it)

File: src/Compiler/RefC/CC.idr [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     let sharedFlag = if asShared then "-shared " else ""
+
+     let runcc = cc ++ " " ++ sharedFlag ++ objectFile ++
+                       " -o " ++ outFile ++ " " ++
+                       fullprefix_dir dirs "lib" </> "libidris2_support.a" ++ " " ++
+                       "-lidris2_refc " ++
+                       "-L" ++ fullprefix_dir dirs "refc " ++
+                       clibdirs (lib_dirs dirs)
+
+     coreLift $ putStrLn runcc
```
Comment (gallais) [inline src/Compiler/RefC/CC.idr:66]: Same remark wrt `putStrLn` vs `log` here.

File: src/Compiler/RefC/RefC.idr [hunk 3]
```diff
@@ -1044,27 +1031,18 @@ executeExpr c _ tm
...
+                   -> Core (Maybe String)
+generateCSourceFile defs outn =
+  do _ <- newRef ArgCounter 0
      _ <- newRef FunctionDefinitions []
      _ <- newRef TemporaryVariableTracker []
      _ <- newRef OutfileText DList.Nil
      _ <- newRef ExternalLibs []
      _ <- newRef IndentLevel 0
-     traverse_ (\(n, d) => createCFunctions n d) defs
+     _ <- traverse (\(n, d) => createCFunctions n d) defs
```
Comment (gallais) [inline src/Compiler/RefC/RefC.idr:1045]: What's the reason behind this change?

File: src/Compiler/RefC/RefC.idr [hunk 4]
```diff
@@ -1073,34 +1051,29 @@ compileExpr ANF c _ outputDir tm outfile =
...
+     let outexec = outputDir </> outfile
 
-     coreLift $ putStrLn runccobj
-     0 <- coreLift $ system runccobj
-       | _ => pure Nothing
-     coreLift $ putStrLn runcc
-     0 <- coreLift $ system runcc
-       | _ => pure Nothing
-     pure (Just outexec)
+     _ <- coreLift $ mkdirAll outputDir
```
Comment (gallais) [inline src/Compiler/RefC/RefC.idr:1069]: Same question here about changing a call to `coreLift_` to `_ <- coreLift`.

File: src/Compiler/RefC/RefC.idr [hunk 5]
```diff
@@ -1073,34 +1051,29 @@ compileExpr ANF c _ outputDir tm outfile =
      coreLift_ $ writeFile outn code
      coreLift_ $ putStrLn $ "Generated C file " ++ outn
```
Comment (gallais) [inline src/Compiler/RefC/RefC.idr:1052]: Same comment about `putStrLn` vs. `log`

File: src/Compiler/RefC/CC.idr (hunk 1 above)
Comment (madman-bob) [reply src/Compiler/RefC/CC.idr:41]: This line is unchanged from the original code, just moved. I'm happy to change it if that's the right thing to do, but I don't know why this line is as it is.

File: src/Compiler/RefC/RefC.idr (hunk 3 above)
Comment (madman-bob) [reply src/Compiler/RefC/RefC.idr:1045]: This change comes from merging `master`. You'd have to ask whoever made the original commit.

File: src/Compiler/RefC/RefC.idr (hunk 4 above)
Comment (madman-bob) [reply src/Compiler/RefC/RefC.idr:1069]: Same response.

File: src/Compiler/RefC/RefC.idr (hunk 3 above)
Comment (ohad) [reply src/Compiler/RefC/RefC.idr:1045]: `Core.Core.traverse` in this case is `Core (List Unit)`, and it seems that to sequence something it needs to return `Core Unit`.

File: src/Compiler/RefC/RefC.idr (hunk 4 above)
Comment (ohad) [reply src/Compiler/RefC/RefC.idr:1069]: Same response.

File: src/Compiler/RefC/RefC.idr (hunk 3 above)
Comment (gallais) [reply src/Compiler/RefC/RefC.idr:1045]: Right so I suspect that the conflict was resolved the wrong way.

File: src/Compiler/RefC/CC.idr (hunk 1 above)
Comment (gallais) [reply src/Compiler/RefC/CC.idr:41]: Right but as we are working on that part of the codebase, that's an opportunity
to clean it up a bit as we go. Debugging information should be communicated using
`log` rather than `putStrLn`. This way it can be turned off if we don't need the
info.

File: src/Compiler/RefC/RefC.idr (hunk 3 above)
Comment (ohad) [reply src/Compiler/RefC/RefC.idr:1045]: I see. Let me give it a go on a fresh branch.

Comment (ohad) [reply src/Compiler/RefC/RefC.idr:1045]: No, manually inserting those changes on a fresh branch gives the same error. Let's chat.

File: src/Compiler/RefC/RefC.idr [hunk 6]
```diff
@@ -1024,7 +1026,7 @@ footer = do
 export
 executeExpr : Ref Ctxt Defs -> (execDir : String) -> ClosedTerm -> Core ()
 executeExpr c _ tm
-    = do coreLift_ $ putStrLn "Execute expression not yet implemented for refc"
+    = do log "compiler.refc" 0 "Execute expression not yet implemented for refc"
```
Comment (gallais) [inline src/Compiler/RefC/RefC.idr:1029]: I think this one is a valid `putStrLn`: it's an error message displayed upon failure.

Comment (madman-bob) [reply src/Compiler/RefC/RefC.idr]: I'll take your word for it. Updated
